requests-for-comments/2021-10-05-group-analytics.md

```diff
@@ -0,0 +1,218 @@
+
+### Group properties API
+
+We will expose a new API for creating and adding properties groups. This API can be exposed on the client level, e.g.
+
+```python
+    import posthog
+
+    posthog.groups.add_properties('organization', 'org:5', { name: 'Swedish Bakery', subscription_level: 'gold' })
+```
+
+There will not be any functionality for merging two groups, as that behavior is user-specific.
+
+### Capture API
+
+Capturing events in client libraries will allow users to pass group metadata alongside with an event.
+
+Here’s how it would look like in python:
+
+```python
+    import posthog
+
+    posthog.capture(
+      user_id,
+      'purchase'
+      { plan: 'silver' },
+      groups={
+        'organization': 'org:6'
+      }
+    )
+```
+
+`groups` metadata will be sent to us under event properties under new the `$groups` key.
+
+Our data ingestion in plugin-server will perform the neccessary mappings from group ids to column indexes or other keys as needed.
+
+
+### In-App: filtering by group properties in insights
+
+In-app, users will be able to do analytics by groups and filter by them.
+
+To give an example:
+
+- Throughout insights, we’ll allow filtering insights with group properties. This will be implemented as a new tab per group type in the filters modal.
+- In trends, users can also select “unique organizations” after first capturing events for an “organization”
+- In funnels, users can select to analyze by a group type (e.g. “instance”). Different users within the group can have completed different steps of the funnel.
+- In retention, users can select “organization” over “user” for the unit to analyze retention by
+- And so on…
+
+
+### In-App: profile pages for groups
+
+We’ll also expose pages for different groups in-app, with:
+
+
+1. Listing all your different group types
+2. Listing and allowing searching all different groups of a type (e.g. organizations)
+3. Group page with
+    1. Group properties
+    2. Events from that group
+    3. Related users and groups (calculated via users’ other events)
+
+Initially, the design and behavior will be taken from the person pages and iterated later on.
+
+
+### Autocapture & posthog-js
+
+To make autocapture work with groups and to make posthog-js usage a bit simpler, we can provide the following API:
+
+```js
+    // Normal posthog usage
+    posthog.identify(userId)
+    // New proposed method, to be called right after .identify
+    posthog.group('instance', 'posthog.yourapp.com', { instance_prop: 5})
+```
+
+After calling this method, we will automatically associate events from that session with that group until either:
+
+- `posthog.group` is called with the same group type
+- `posthog.reset()` is called.
+### Feature flags
+
+Users should be able to add group-level property filters to feature flags and be able to ensure rollouts at a group level (?).
```

Comment:
Another (separate or complimentary) approach here may be support for feature flags created _against_ group types instead of users. I.e. you specify @ flag creation what kinds of groups this flag is being rolled to.